api/lib/domain/models/CertificationCandidate.js

```diff
@@ -302,6 +243,14 @@ class CertificationCandidate {
   convertExtraTimePercentageToDecimal() {
     this.extraTimePercentage = this.extraTimePercentage / 100;
   }
+
+  /**
+   * @param {Subscription} subscription
+   */
+  addSubscription(subscription) {
+    this.subscriptions = this.subscriptions.filter(({ type }) => subscription.type !== type);
```

Comment:
Pareil sur ce `filter`, je n'arrive pas à percevoir dans quel cas on en a besoin 🤔 ? Ne serait-ce pas une sécurité en trop ?

Reply:
A rediscuter ! C'est pour éviter des doublons, mais possible que ça n'ait pas de sens

Reply:
Pour moi ce n'est pas une sécurité, c'est une importante règle métier certif.

En fait ici, le filter est ce qui déporte la responsabilité au domain model d'être "sain". Le model, via `addSubscription` est garant pour tous les consommateurs que tu peux jamais (grace au filter) arriver dans une situation où tu permets d'avoir deux souscriptions complémentaires (ou deux coeur), c'est une règle métier assez importante chez certif.

Ce n'est ainsi plus aux X consommateurs de faire la gymnastique où ils doivent s'assurer qu'ils vont pas corrompre les règles métiers ; c'est le model qui n'est pas anémique qui assure cela.

Exemple: 
* Imaginons que tu fais pas gaffe et que le usecase de la modale fait deux fois un `addSubscription` CORE (en gros tu as un bug dans une boucle ou un concat d'array, ça arrive)
* Dans cette version du code actuel : aucun soucis, ton bug va pas corrompre la BDD et pas besoin d'ajouter des sécurité, tes règles métiers sont assurées par le `addSubscription`
  * Rappelons d'ailleurs que le `validate` est pas toujours appelé aujourd'hui (c'est pas top, mais c'est ainsi pour ce model), donc on peut pas faire confiance à Joi pour éviter une corruption des règles
* Sans le filter, tu aurais deux items CORE dans ton array, et du coup les options suivantes pour éviter cela :
  * le repository soit il commence à ajouter des garde-fous (c'est pas bien, tu commences à mettre du métier dans tes repository)
  * tu pries pour que tes tests le voient (et bon, on a vu que nos tests sont faillibles clairement)
  * tu mets un garde-fou dans ton usecase (et du coup dans **tous** les usecases faut y penser), en gros c'est le usecase qui porte la règle métier et point de vue théorie DDD c'est pas forcèment bien, en + de savoir que cette règle existe

N'étant pas non plus un expert DDD, j'ai cherché des infos dans le bouquin pour vérifier cela, et ça me semble cohérent :)

> What’s a Domain Model?
It’s a software model of the very specific business domain you are working in. Often it’s implemented as an object model, where those objects have both data and behavior with literal and accurate business meaning.

> The Language of a team in an explicit Bounded Context expressed as a domain model adds true business value and gives us certainty that we are implementing the correct software. Even from a technical standpoint, it helps us create better models, ones with more potent behaviors, that are pure and less error prone.

Reply:
A noter que on aurait pu appeler la méthode éventuellement `updateSubscriptions`, pour clarifier le comportement de cette fonction

Reply:
Merci pour toutes les explications, c'est très clair 🙏